File: 1079-letter-tile-possibilities/1079-letter-tile-possibilities.py
class Solution:
    def _find_sequences(self, char_cnt: Dict[chr, int]) -> int:
        ans = 0

        # Try using each available character
        for c in char_cnt:
            if char_cnt[c] == 0:
                continue
            
            # Add current character and recurse
            ans += 1
            char_cnt[c] -= 1
            ans += self._find_sequences(char_cnt)
            char_cnt[c] += 1

        return ans

    def numTilePossibilities(self, tiles: str) -> int:
        # Track frequency of each uppercase letter (A-Z)
        char_cnt = dict()
        for c in tiles:
            char_cnt[c] = char_cnt.get(c, 0) + 1
        
        # Find all possible sequences using character frequencies
        return self._find_sequences(char_cnt)

File: 1079-letter-tile-possibilities/1079-letter-tile-possibilities.cpp
class Solution {
private:
    int findSequences(unordered_map<char, int>& charCount){
        int total = 0;

        // Try using each available character
        for (auto& [ch, freq] : charCount){
            if (charCount[ch] == 0) continue;

            // Add current character and recurse
            ++total;
            --charCount[ch];
            total += findSequences(charCount);
            ++charCount[ch];
        }

        return total;
    }

public:
    int numTilePossibilities(string tiles) {
        // Track frequency of each characters
        unordered_map<char, int> charCount;
        for (const char& c : tiles) ++charCount[c];

        // Find all possible sequences using character frequencies
        return findSequences(charCount);
    }
};